<html>
  <head>
    <link rel = "stylesheet" href = "../default.css">
  </head>
  <body>
    <h1>Polymorphic Edge Detection (PED)<font size = "4"> - An efficient polymorphism detector for NGS data</font></h1>
    <a href = "../index.html">[Link to Top Page]</a>
    <h2>Example of SARS-CoV-2 Mutation Detection</h2>
    <a href = "SARS-CoV-2.png" target = "_blank"><img src = "SARS-CoV-2.png" width = "30%"></a> Click to enlarge.<br>
    <ul>
    <li><b>Install PED into linux machine (<a href = "http://ubuntu.com/" target = "_blank">Ubuntu</a>),</b><br>
    $ sudo apt update<br>
    $ sudo apt upgrade<br>
    $ sudo apt install curl<br>
    $ sudo apt install git<br>
    $ git clone https://github.com/akiomiyao/ped.git<br>

    <li><b>Install and setup fastq-dump.</b><br>
    <a href = "https://akiomiyao.github.io/ped/sratoolkit/index.html" target = "_blank">Details for setup is Here.</a>
    <li><b>Download fastq sequence data from NCBI sequence read archive</b><br>
    $ cd ped<br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11513115" target = "_blank">SRR11513115</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11513116" target = "_blank">SRR11513116</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11513117" target = "_blank">SRR11513117</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11513118" target = "_blank">SRR11513118</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11542243" target = "_blank">SRR11542243</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11542244" target = "_blank">SRR11542244</a><br>
    <li><b>Detection of polymorphism</b><br>
    $ perl ped.pl target=SRR11513115,ref=SARS-CoV-2<br>
    $ perl ped.pl target=SRR11513116,ref=SARS-CoV-2<br>
    $ perl ped.pl target=SRR11513117,ref=SARS-CoV-2<br>
    $ perl ped.pl target=SRR11513118,ref=SARS-CoV-2<br>
    $ perl ped.pl target=SRR11542243,ref=SARS-CoV-2<br>
    $ perl ped.pl target=SRR11542244,ref=SARS-CoV-2<br>
    Currently, fastq files of Illumina and Oxford nanopore are supported.
    <li><b>For Docker users</b><br>
    $ sudo apt install docker<br>
    $ sudo apt install docker.io<br>
    Example of analysis,<br>
    $ sudo docker pull akiomiyao/ped<br>
    $ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl download.pl accession=SRR11513115,wd=/work <br>
    $ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl ped.pl target=SRR11513115,ref=SARS-CoV-2,wd=/work <br>
    Copy and paste commands described above to your terminal window, change values of accession, target and run.<br>
    Both the ped.pl script and the docker container can be run on Google Cloud Platform and Amazon Web Services.
    <li><b>For private data</b><br>
    $ mkdir your_data1 (target name for ped.pl is the directory name)<br>
    $ mkdir your_data1/read<br>
    $ cp ~/your_data_directory/foo.fastq your_data1/read (compressed file with gz, bzip2 and xz extentions can be used)<br>
    $ perl ped.pl target=your_data1,ref=SARS-CoV-2<br>
    or<br>
    $ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl ped.pl target=your_data1,ref=SARS-CoV-2,wd=/work<br>
    <li><b>About clipping</b><br>
    Fastq sequence of RT-PCR products is not fixed length. Since PED requires fixed length sequences, clipping of sequences is required.<br>
    The new ped.pl determines optimal clipping length. If you want to force the clipping length, add clipping option.<br>
    For example,<br>
    $ perl ped.pl target=SRR11513115,ref=SARS-CoV-2,clipping=146<br>
    To know the optimal clipping length, run check_length.pl.<br> 
    $ perl check_length.pl target=SRR11513115<br>
    'check_length.pl' outputs read counts in each length.<br>
    Choose the optimal length for clipping.<br>
    NCBI also shows the optimal length. See the metadata of each run.<br>
    For data of Oxford nanopore, clipping=100 or 150 is recommended.
    <li><b>Viewing mutations with Integrative Genomics Viewer</b><br>
    Integrative Genomics Viewer (IVG) is a product of Broad Institute.<br>
Web application of IVG <a href = "https://igv.org/app/" target = "_blank">https://igv.org/app/</a> is useful.<br>
<a href = "igv-web.png" target = "_blank"><img src = "igv-web.png" height = "30%"></a> Click to large. <br>
      
      Select Genome to Sars-CoV-2 (ASM985889v3) from the pull down menu. <br>
      Download pre-analyzed vcf files linked below. <br>
      Import files by selecting 'Local file' in 'Tracks' pull down menu.<br><br>
      
    
    If you want to view by your computer, download and install IVG from <a href = "http://software.broadinstitute.org/software/igv/download" target = "_blank">http://software.broadinstitute.org/software/igv/download</a>.<br>
    Click Genomes tab of IVG and load "SARS-CoV-2.fasta" and "SARS-CoV-2.gff" files.<br>
    Click File tab and load "ped/SRR11513115/SRR11513115.vcf" and other vcf files or drop files on the main window.<br>
    <li><b>Check mutation in bidirectional alignments</b><br>
    In the ped directory,<br>
$ perl search.pl target=target,chr=chromosome,pos=position<br><br>
For example,<br>
$ perl search.pl target=SRR11542244,chr=NC_045512.2,pos=5872<br><br>
or<br><br>
$ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl search.pl target=SRR11542244,chr=NC_045512.2,pos=5872,wd=/work <br><br>
A part of output is<br>

<pre># NC_045512.2 5872 NC_045512.2 5870 f deletion -1
CTTACAAAGTCCTCAGAATA GTTACTTATAAATTGGATGG 0
                                                  ChrNC_045512.2 5872
                                                  |
CTTACAAAGTCCTCAGAATACAAAGGTCCTATTACGGATGTTTTCTACAAAGAAAACAGTTACACAACAACCATAAAACCAGTTACTTATAAATTGGATG
||||||||||||||||||||||||||||||||||||||||||||||||||  |||    |     |  | |   ||| |   |   |   || | |   |
CTTACAAAGTCCTCAGAATACAAAGGTCCTATTACGGATGTTTTCTACAAGAAAACAGTTACACAACAACCATAAAACCAGTTACTTATAAATTGGATGG
 |   ||   |     |    || |  |   |   |    |||     ||||||||||||||||||||||||||||||||||||||||||||||||||||
TTACAAAGTCCTCAGAATACAAAGGTCCTATTACGGATGTTTTCTACAAAGAAAACAGTTACACAACAACCATAAAACCAGTTACTTATAAATTGGATGG
                                               |
                                               ChrNC_045512.2 5870
</pre>
    The position for indel in vcf file is one base before the junction position. On the other hand, alignment of ped shows the position of junction.<br>
    In the case of no alignment, search with vcf position plus one.<br>
    <li><b>Files</b><br>
    <a href = "SARS-CoV-2.fasta" target = "_blank">SARS-CoV-2.fasta</a> From NCBI<br>
    <a href = "SARS-CoV-2.gff" target = "_blank">SARS-CoV-2.gff</a> From NCBI<br>
    <a href = "SRR11542243.count.vcf" target = "_blank">SRR11542243.count.vcf</a> Sample from University Malaya<br>
    <a href = "SRR11542244.count.vcf" target = "_blank">SRR11542244.count.vcf</a> Sample from University Malaya<br>
    <a href = "SRR11513115.count.vcf" target = "_blank">SRR11513115.count.vcf</a> Sample from Utah Public Health Laboratory<br>
    <a href = "SRR11513116.count.vcf" target = "_blank">SRR11513116.count.vcf</a> Sample from Utah Public Health Laboratory<br>
    <a href = "SRR11513117.count.vcf" target = "_blank">SRR11513117.count.vcf</a> Sample from Utah Public Health Laboratory<br>
    <a href = "SRR11513118.count.vcf" target = "_blank">SRR11513118.count.vcf</a> Sample from Utah Public Health Laboratory<br>
    <li><b>Comments</b><br>
    Run time is about 2 min for each accession using Intel Core i5-4210M CPU @ 2.60 GHz (2 cores, 4 threads), 8GB memory.<br><br>
    Reads from Illumina platform returns fine results. On the other hand, results from nanopore contains a lot of false positives.<br>
    Higher threshold may be required for analysis of nanopore data.<br><br>
    PED outputs three vcf format files.<br>
    'target.vcf' is selected data for IGV.<br>
    'target.full.vcf' is full output before the selection.<br>
    'target.count.vcf' is from raw data before the verification.<br>
    Some mutations have been filtered out by the verification step, because of too much coverage of short reads.<br>
    In addition, two mutations in nearby position (within the clipping length) but on different RNA strand (from mixture of viruses) will be filtered out by verification step.<br>
    Usage of count.vcf files from raw count data before verification step is recommended. 
    <li><b>Acknowledgement</b><br>
    I thank Dr Ilkka Havukkala for testing the PED update for SARS-CoV-2 and advice about integration of IVG.<br>
    <li><b>Citing PED</b><br>
Cite this article as: Polymorphic edge detection (PED): two efficient methods of polymorphism detection from next-generation sequencing data<br>
Akio Miyao, Jianyu Song Kiyomiya, Keiko Iida, Koji Doi, Hiroshi Yasue<br>
BMC Bioinformatics. 2019 20(1):362.<br>
URL <a href = "https://doi.org/10.1186/s12859-019-2955-6" target = "_blank">https://doi.org/10.1186/s12859-019-2955-6</a>
    <li><b>Author</b><br>
    MIYAO Akio (miyao@affrc.go.jp)
    </ul>
  </body>
</html>
