<html>
  <head>
    <link rel = "stylesheet" href = "../default.css">
  </head>
  <body>
    <h1>Polymorphic Edge Detection (PED)<font size = "4"> - An efficent polymorphism detector for NGS data</font></h1>
    <a href = "../index.html">[Link to Top Page]</a>
    <h2>Example of COVID19 Mutation Detection</h2>
    <a href = "covid19.png" target = "_blank"><img src = "covid19.png" width = "30%"></a> Click to large.<br>
    <ul>
    <li><b>For Install linux machine (<a href = "http://ubuntu.com/" target = "_blank">Ubuntu</a>),</b><br>
    $ sudo apt-get update<br>
    $ sudo apt install curl<br>
    $ sudo apt install git<br>
    $ sudo apt install sra-toolkit<br>
    $ git clone https://github.com/akiomiyao/ped.git<br>
    <li><b>Download fastq sequence data from NCBI sequence read archive</b><br>
    $ cd ped<br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11513115" target = "_blank">SRR11513115</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11513116" target = "_blank">SRR11513116</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11513117" target = "_blank">SRR11513117</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11513118" target = "_blank">SRR11513118</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11542243" target = "_blank">SRR11542243</a><br>
    $ perl download.pl accession=<a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR11542244" target = "_blank">SRR11542244</a><br>
    <li><b>Detection of polymorphism</b><br>
    $ perl ped.pl target=SRR11513115,ref=COVID19,clipping=146<br>
    $ perl ped.pl target=SRR11513116,ref=COVID19,clipping=146<br>
    $ perl ped.pl target=SRR11513117,ref=COVID19,clipping=146<br>
    $ perl ped.pl target=SRR11513118,ref=COVID19,clipping=143<br>
    $ perl ped.pl target=SRR11542243,ref=COVID19,clipping=100<br>
    $ perl ped.pl target=SRR11542244,ref=COVID19,clipping=100<br>
    Currently, fastq files of Illmina and Oxford nanopore are supported.
    <li><b>Size of clipping</b><br>
    Fastq sequence of RT-PCR products is not fixed length. Since PED requires fixed length sequences, clipping of sequences is required.<br>
    $ perl check_length.pl target=SRR11513115<br><br>
    outputs read counts in each length. Choose the optimal length for clipping.<br>
    NCBI also shows the optimal length. See the metadata of each run.<br>
    For data of Oxford nanopore, clipping=100 or 150 is recommended.
    <li><b>Viewing mutations with Integrative Genomics Viewer</b><br>
    Integrative Genomics Viewer (IVG) is a product of Broad Institute.<br>
    Download and install IVG from <a href = "http://software.broadinstitute.org/software/igv/download" target = "_blank">http://software.broadinstitute.org/software/igv/download</a>.<br>
    Fasta file of reference is saved in "ped/COVID19/COVID19.fasta".<br>
    Click Genomes tab of IVG and load "COVID19.fasta" and "COVID19.gff" files.<br>
    Click File tab and load "ped/SRR11513115/SRR11513115.vcf" and other vcf files or drop files on the main window.<br>
    <li><b>Check mutation in bidirectional alignments</b><br>
    In the ped directory,<br>
$ perl search.pl target=target,chr=chromosome,pos=position<br><br>
For example,<br>
$ perl search.pl target=SRR11542244,chr=1,pos=5872<br><br>
A part of output is<br>

<pre># 1 5872 1 5870 f deletion -1
CTTACAAAGTCCTCAGAATA GTTACTTATAAATTGGATGG 0
                                                  Chr1 5872
                                                  |
CTTACAAAGTCCTCAGAATACAAAGGTCCTATTACGGATGTTTTCTACAAAGAAAACAGTTACACAACAACCATAAAACCAGTTACTTATAAATTGGATG
||||||||||||||||||||||||||||||||||||||||||||||||||  |||    |     |  | |   ||| |   |   |   || | |   |
CTTACAAAGTCCTCAGAATACAAAGGTCCTATTACGGATGTTTTCTACAAGAAAACAGTTACACAACAACCATAAAACCAGTTACTTATAAATTGGATGG
 |   ||   |     |    || |  |   |   |    |||     ||||||||||||||||||||||||||||||||||||||||||||||||||||
TTACAAAGTCCTCAGAATACAAAGGTCCTATTACGGATGTTTTCTACAAAGAAAACAGTTACACAACAACCATAAAACCAGTTACTTATAAATTGGATGG
                                               |
                                               Chr1 5870
</pre>
Because COVID-19 has single RNA genome, the name of chromosome is renumbered to 1 for the PED analysis.

</pre>
    <li><b>Files</b><br>
    <a href = "COVID19.fasta" target = "_blank">COVID19.fasta</a> From NCBI and modified for PED.<br>
    <a href = "COVID19.gff" target = "_blank">COVID19.gff</a> From NCBI and modified for PED.<br>
    <a href = "SRR11542243.vcf" target = "_blank">SRR11542243.vcf</a> Sample from University Malaya<br>
    <a href = "SRR11542244.vcf" target = "_blank">SRR11542244.vcf</a> Sample from University Malaya<br>
    <a href = "SRR11513115.vcf" target = "_blank">SRR11513115.vcf</a> Sample from Utah Public Health Laboratory<br>
    <a href = "SRR11513116.vcf" target = "_blank">SRR11513116.vcf</a> Sample from Utah Public Health Laboratory<br>
    <a href = "SRR11513117.vcf" target = "_blank">SRR11513117.vcf</a> Sample from Utah Public Health Laboratory<br>
    <a href = "SRR11513118.vcf" target = "_blank">SRR11513118.vcf</a> Sample from Utah Public Health Laboratory<br>
    <li><b>Comments</b><br>
    Run time is about 20 min for each accession using Intec Core i5-4210M CPU @ 2.60 GHz (2 cores, 4 threads), 8GB memory.<br>
    Reads from Illumina platform returns fine results. On the other hand, results from nanopore contains a lot of false positives.<br>
    Higher threshold may be required for analysis of nanopore data.
    <li><b>Acknowledgement</b><br>
    I thank Dr Ilkka Havukkala for test the PED update for COVID-19 and advice about integration of IVG.<br>
    <li><b>Citing PED</b><br>
Cite this article as: Polymorphic edge detection (PED): two efficient methods of polymorphism detection from next-generation sequencing data<br>
Akio Miyao, Jianyu Song Kiyomiya, Keiko Iida, Koji Doi, Hiroshi Yasue<br>
BMC Bioinformatics. 2019 20(1):362.<br>
URL <a href = "https://doi.org/10.1186/s12859-019-2955-6" target = "_blank">https://doi.org/10.1186/s12859-019-2955-6</a>
    <li><b>Author</b><br>
    MIYAO Akio (miyao@affrc.go.jp)
    </ul>
  </body>
</html>
