<html>
  <head>
    <meta charset = "utf-8">
    <link rel = "stylesheet" href = "../default.css">
  <head>
    <body>
      <h1>Detection of polymorphisms from <i>Escherichia coli</i> NGS by PED</h1>
      <a href = "../index.html">[Link to top page]</a><br><br>
      <b>In this section, install of Linux(Ubuntu), setup of PED and browsing of detected polymorphisms using <a href = "https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP261658" target = "_blank">NGS of mutant strains of <i>Escherichia coli</i> str. K-12 substr. MG1655</a> are described.</b><br><br>
An example of genome view by IGV. Click to large.
      <ul>
      <li>Click to large.<br><a href = "Ecoli.png" target = "_blank"><img src = "Ecoli.png" height = "30%"></a>
<a href = "RpoB-P153L.png" target = "_blank"><img src = "RpoB-P153L.png" height = "30%"></a>
<li><b>Indels can be detect exactly,</b> as well as SNPs.
<li><b>Nucleotide sequences of primer pair</b> to amplify target mutations are also shown.
      </ul>
Strain CAG67106 (SRR11791057) is a suppressor mutant of <i>mreB</i> mutation.<br>
Mutation at 4,181,702 is C to T substitution (P153L) in the <i>RpoB</i> gene.<br>
      <h2>Install of Ubuntu</h2>
      <ul>
      <li>Download Ubuntu 20.4 LTS from <a href = "https://ubuntu.com/download" target = "_blank">Ubuntu download site</a>.
      <li><a href = "https://ubuntu.com/tutorials/tutorial-install-ubuntu-desktop" target = "_blank">Ubuntu install tutorial</a> will be helpful for installing.<br>
      Because PED uses maximum CPU power, exclusive use for PED analysis is recommended.<br>
      PED supports multithread environment. CPU with 4 or 8 core is recommended.
      </ul>
      <h2>Setup and analysis of PED</h2>
      <ul>
      <li>Installing software required by PED
      <pre>$ sudo apt update
$ sudo apt upgrade
$ sudo apt install curl
$ sudo apt install git
$ git clone https://github.com/akiomiyao/ped.git</pre>
      <li>To download sequence data, fastq-dump from NCBI is required.<br>
Tool kit can be download from
<a href = "https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software" target = "_blank">https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software</a><br>
Details of setup fastq-dump is described in
<a href = "https://akiomiyao.github.io/ped/sratoolkit/index.html" target = "_blank">https://akiomiyao.github.io/ped/sratoolkit/index.html</a>
      <li>Download fastq files from NCBI Sequence read archive site.
      <pre>$ cd ped
$ git pull
$ perl download.pl accession=SRR11791057
$ perl download.pl accession=SRR11791061
</pre>
Scripts are updated by git pull command.<br>
Sometimes error message will appear during download. The script will try to reconnect. Check increasing size of downloading files.<br>
<li>Detection of polymorphisms in NGS sequences on <i>Escherichia coli</i> reference genome.
<pre>$ perl ped.pl target=SRR11791057,control=SRR11791061,ref=Ecoli
</pre>
Since Ecoli reference file is already configured, data are automatically downloaded and calculated.<br>
When all processes are done, vcf files will be created in target directory.<br>
Common mutations in the reference sequence with both target and control strains will be eliminated.<br>
Sequences of primer pair are also outputted into vcf file.
      </ul>
      <h2>For Docker users</h2>
      <ul>
      <li>Install of Docker<br>
<pre>$ sudo apt install docker
$ sudo apt install docker.io</pre>
      <li>Example of analysis,<br>
<pre>$ sudo docker pull akiomiyao/ped
$ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl download.pl accession=SRR11791057,wd=/work
$ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl download.pl accession=SRR11791061,wd=/work
$ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl check_length.pl accession=SRR7967955,wd=/work
$ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl ped.pl target=SRR11791057,control=SRR11791061,ref=Ecoli,wd=/work</pre>
      Copy and paste commands described above to your terminal window, change values of accession, target and clipping, and run.<br>
      Both the ped.pl script and the docker container can be run on Google Cloud Platform and Amazon Web Services.
      </ul>
      <h2>Visualization of detected polymorphisms using IVG</h2>
      <ul>
      <li>Integrative Genomics Viewer (IVG) is a genome browser made in Broad Institute.<br>
Web application of IVG <a href = "https://igv.org/app/" target = "_blank">https://igv.org/app/</a> is useful.<br>
<a href = "RpoB-P153L_igv1.png" target = "_blank"><img src = "RpoB-P153L_igv1.png" height = "10%"></a>
<a href = "RpoB-P153L_igv2.png" target = "_blank"><img src = "RpoB-P153L_igv2.png" height = "10%"></a> Click to large.
      <ul>
      <li>Download fasta file of reference genome <a href = "https://tos.nias.affrc.go.jp/~miyao/pub/20200707BSUB/Ecoli.fasta" target = "_blank">Ecoli.fasta</a> and the index file <a href = "https://tos.nias.affrc.go.jp/~miyao/pub/20200707BSUB/Ecoli.fasta.fai" target = "_blank">Ecoli.fasta.fai</a>.
      <li>Select 'Local File ...' in pull down menu of 'Genome' and submit both Ecoli.fasta and Ecoli.fasta.fai files.
      <li>Download gff file. <a href = "https://tos.nias.affrc.go.jp/~miyao/pub/20200707BSUB/GCF_000005845.2_ASM584v2_genomic.gff.gz" target = "_blank">GCF_000005845.2_ASM584v2_genomic.gff.gz</a>
      <li>Select 'Local File ...' in pull down menu of 'Tracks' and submit GCF_000005845.2_ASM584v2_genomic.gff.gz file.
      <li>Download pre-analyzed vcf file <a href = "SRR11791057.vcf" target = "_blank">SRR11791057.vcf</a>.
      <li>Import files by selecting 'Local file' in 'Tracks' pull down menu.
      <li>Click '+' to zoom up. Vertical lines of mutations will be shown.
      </ul>
      <li>Or download from <a href = "http://software.broadinstitute.org/software/igv/download" target = "_blank">http://software.broadinstitute.org/software/igv/download</a>.<br>
Drag vcf files on main display panel of IVG, data will be imported.<br>
Click at interested position, zoom up the region.<br>
Mouse over at vertical line of polymorphism, detailed information will be shown.<br>
      </ul>
      <h2>Link of PED</h2>
      <ul>
       <li>Github: <a href = "https://github.com/akiomiyao/ped/" "_blank">https://github.com/akiomiyao/ped/</a>
       <li>Docker: sudo docker pull akiomiyao/ped<br>
       <li>Paper: <a href = "https://doi.org/10.1186/s12859-019-2955-6" target = "_blank">Polymorphic edge detection (PED): two efficient methods of polymorphism detection from next-generation sequencing data (2019) BMC Bioinformatics 20(1):362</a>
      </ul> 
      <h2>Contact</h2>
MIYAO Akio, Ph.D (miyao@affrc.go.jp) (Family name is MIYAO)<br>
Institute of Crop Science / NARO<br>
2-1-2, Kannondai, Tsukuba, Ibaraki, 305-8518, Japan<br>
    </body>
</html>