<html>
  <head>
    <meta charset = "utf-8">
    <link rel = "stylesheet" href = "../default.css">
  <head>
    <body>
      <h1>Detection of polymorphisms from human NGS by PED</h1>
      <a href = "../index.html">[Link to top page]</a><br><br>
      <b>In this section, install of Linux(Ubuntu), setup of PED and browsing of detected polymorphisms using <a href = "https://www.illumina.com/platinumgenomes.html" target = "_blank">Platinum human sequence data provided by Illumina</a> are described.</b><br><br>
Following is PED analysis with <a href = "https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=ERR194146" target = "_blank">ERR194146</a> and
      <a href = "https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=ERR194147" target = "_blank">ERR194147</a> sequence data.<br>
Project title of short reads: Whole genome sequencing and variant calls for the Coriell CEPH/UTAH 1463 family to create a "platinum" standard comprehensive set for variant calling improvement <a href = "https://www.ncbi.nlm.nih.gov/bioproject/PRJEB3381" target = "_blank">(PRJEB3381)</a>.<br>
      <table>
      <tr><td><a href = "ERR194146_ERR194147.png" target = "_blank"><img src = "ERR194146_ERR194147.png" height = "30%"></a>
<a href = "ERR194146.png" target = "_blank"><img src = "ERR194146.png" height = "30%"></a></td><td><ul><li>An example of genome view by IGV. Click to large.
<li><b>Indels can be detect exactly,</b> as well as SNPs.
<li><b>Nucleotide sequences of primer pair</b> to amplify target mutations are also shown.</ul></td></tr>
      </table>

      <h2>Install of Ubuntu</h2>
      <ul>
      <li>Download Ubuntu 20.4 LTS from <a href = "https://ubuntu.com/download" target = "_blank">Ubuntu download site</a>.
      <li><a href = "https://ubuntu.com/tutorials/tutorial-install-ubuntu-desktop" target = "_blank">Ubuntu install tutorial</a> will be helpful for installing.<br>
      Because PED uses maximum CPU power, exclusive use for PED analysis is recommended.<br>
      PED supports multithread environment. CPU with 4 or 8 core is recommended.
      </ul>
      <h2>Setup and analysis of PED</h2>
      <ul>
      <li>Installing software required by PED
      <pre>$ sudo apt update
$ sudo apt install curl
$ sudo apt install git
$ sudo apt install sra-toolkit
$ git clone https://github.com/akiomiyao/ped.git</pre>
      <li>Download fastq files of the Platinum genome sequence (ERR194146 and ERR194147) from NCBI Sequence read archive site.
      <pre>$ cd ped
$ git pull
$ perl download.pl accession=ERR194146
$ perl download.pl accession=ERR194147</pre>
Scripts are updated by git pull command.<br>
Files of fastq format are downloaded into ERR194146/read and ERR194147/read.<br>
Downloading is slow. It requires one or two days.<br>
Sometimes error message will appear. The script will try to reconnect. Check increasing size of downloading files.<br>
      <li>Polymorphism detection of ERR194146 and ERR194147 sequences on human reference genome hg19.
<pre>$ perl ped.pl target=ERR194146,ref=hg19
$ perl ped.pl target=ERR194147,ref=hg19</pre>
Since hg19 reference file is already configured, data are automatically downloaded and calculated.<br>
When all processes are done, ERR194146.vcf and ERR194147.vcf are created in ERR194146 and ERR194147 directory.<br>
Sequences of primer pair are also outputted into vcf file.
      </ul>
      <h2>For Docker users</h2>
      <ul>
      <li>Set up of Docker<br>
      $ sudo apt install docker<br>
      $ sudo apt install docker.io<br>
      <li>Example of analysis,<br>
      $ sudo docker pull akiomiyao/ped<br>
      $ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl download.pl accession=ERR194146,wd=/work <br>
      $ sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl ped.pl target=ERR194146,ref=hg19,wd=/work <br>
      Copy and paste commands described above to your terminal window, change values of accession, target and clipping, and run.<br>
      Both the ped.pl script and the docker container can be run on Google Cloud Platform and Amazon Web Services.
      </ul>
      <h2>Visualization of detected polymorphisms using IVG</h2>
      <ul>
      <li>Integrative Genomics Viewer (IVG) is a genome browser made in Broad Institute.<br>
Web application of IVG <a href = "https://igv.org/app/" target = "_blank">https://igv.org/app/</a> is useful.<br>
<a href = "igsf3l.png" target = "_blank"><img src = "igsf3l.png" height = "10%"></a>
<a href = "igsf3.png" target = "_blank"><img src = "igsf3.png" height = "10%"></a>
<a href = "ERR194146.png" target = "_blank"><img src = "ERR194146.png" height = "10%"></a> Click to large.
      <ul>
      <li>Download pre-analyzed vcf files linked below.
      <li>Import files by selecting 'Local file' in 'Tracks' pull down menu. Because vcf files are big, it takes minutes. <br>
      </ul>
      <li>Or download from <a href = "http://software.broadinstitute.org/software/igv/download" target = "_blank">http://software.broadinstitute.org/software/igv/download</a>.<br>
Drag ERR194146.vcf and ERR194147.vcf on main display panel of IVG, data will be imported.<br>
Due to big size of vcf data, IVG requests making index. Click OK and create index.<br>
For making index, enough memory (8GB for linux is enough) is required, otherwise your computer will freeze.<br>
If index making is completed but failed to display data, restart the IVG. Data will be shown.<br>
When click the number of chromosome, map and polymorphisms by each chromosome will be shown.<br>
Click at interested position, zoom up the region.<br>
Mouse over at vertical line of polymorphism, detailed information will be shown.<br>
Sequence of primer pair for target polymorphism and PCR product size as well as kind and frequency of polymorphism will also be shown.<br>
      <li><a href = "https://tos.nias.affrc.go.jp/~miyao/pub/20200511AMVT/ERR194146.vcf" target = "_blank">ERR194146.vcf</a><br>
      <a href = "https://tos.nias.affrc.go.jp/~miyao/pub/20200511AMVT/ERR194146.vcf.idx" target = "_blank">ERR194146.vcf.idx</a><br>
      <a href = "https://tos.nias.affrc.go.jp/~miyao/pub/20200511AMVT/ERR194147.vcf" target = "_blank">ERR194147.vcf</a><br>
      <a href = "https://tos.nias.affrc.go.jp/~miyao/pub/20200511AMVT/ERR194147.vcf.idx" target = "_blank">ERR194147.vcf.idx</a><br>
      are already analyzed data.
      </ul>
      <h2>Link of PED</h2>
      <ul>
       <li>Github: <a href = "https://github.com/akiomiyao/ped/" "_blank">https://github.com/akiomiyao/ped/</a>
       <li>Docker: sudo docker pull akiomiyao/ped<br>
&emsp;&emsp;&emsp;&emsp;sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl download.pl accession=ERR194146,wd=/work<br>
&emsp;&emsp;&emsp;&emsp;sudo docker run -w /ped -v `pwd`:/work akiomiyao/ped perl ped.pl target=ERR194146,ref=hg19,wd=/work<br>
<li>Paper: <a href = "https://doi.org/10.1186/s12859-019-2955-6" target = "_blank">Polymorphic edge detection (PED): two efficient methods of polymorphism detection from next-generation sequencing data (2019) BMC Bioinformatics 20(1):362</a>
      </ul> 
      <h2>Contact</h2>
MIYAO Akio, Ph.D (miyao@affrc.go.jp) (Family name is MIYAO)<br>
Institute of Crop Science / NARO<br>
2-1-2, Kannondai, Tsukuba, Ibaraki, 305-8518, Japan<br>
    </body>
</html>